README.md
# Submission for 18CSC302J - Computer Networks
This is a **fully featured chess app** written purely in Python using Pygame Library.

![main image](screenshots/main.jpg)

Click [here](screenshots/screenshots.md) to see more few screenshots of My-PyChess in action!

Any bug-reports, suggestions or questions, you can leave it in the github issues section.
If you want to directly communicate with me, you can mail me: itsankith26@gmail.com

The My-PyChess project is available under MIT License. The MIT Licence applies to all the resources I have created in this project. This includes all the python files and text files. But some resources(images, sounds and font file) are not created by me, I have downloaded these from the internet. I have given credits to the authors of these resources in [this file](res/CREDITS.txt). All these resources maintain the original licenses that the authors have leased them under (These licenses permit my use of the respective resources in this project).

## Getting started, with python source code(All OS support)
- Make sure you have **python** and **pygame** installed and working.
- Clone this repository (or download zip file and extract it).
- Then, run the **pychess.py** file. Trying to run any other file will not run the game.
- Those who are running older versions of this app, can upgrade to the latest version by the same method.

## Getting started, by downloading executable(Windows OS only)
- If you are only interested in chess and not python programming, head to releases section on github. 
- There, you will find executable packages under the respective versions. 
- Download the My-PyChess-win-exe.zip from the latest version, extract the folder into your computer.
- Run My-PyChess.exe
- Those who are running older versions of this app, can upgrade to the latest version by the same method.

Interested in python game development with pygame and want to learn from this applicaton - I have released a [lite implementation](https://github.com/ankith26/My-PyChess-lite/) of My-PyChess, that focuses just on chess programming - free from all the code for menus, singleplayer, online etc. 

## Features
- Clean GUI with a lot of menus for ease of use.
- Allows users to make only valid move, does not allow users to make moves that puts the user's king at check.
- Detects check, checkmate, stalemate and informs user.
- Supports things like castling, pawn promotion, enpassent etc.
- Supports saving and loading games.
- Has single player mode with two different types, levels and ability to play against the stockfish chess engine.
- Has online gamemode, play chess with anyone in the world.
- Has a chess game timer.
- Has a preference menu where you can customize the app to meet your preferences.
- Has a chess howto, about menu and stockfish install/configure menu to make things easy for users.

## What's new in this Version 3.2
- Added a Back-Button, to go back to the previous menu. In older versions, the quit button was used for this purpose, but from now on a dedicated button to go back is there in the top-right corner. The quit button will now be used only to exit the app.
- Added a game timer to multiplayer mode, with a new menu to setup the timer.
- Fixed bugs and made many additions and improvements to the client-server in chess online gamemode.
- Made optimisations to core chess module and gui module.
- Added a chess hotwto.
- Minor improvements to game sounds and textbox.
- Upgraded preference menu and made the loadgame interface more robust.
- Several other minor changes and improvements made.

## Highlights of v3.x, the latest major release.
- The code was revamped and restructured, fixed bugs and made **major performance improvements**.
- **Online play** was added. This features a online lobby and support upto 10 people to play chess.
- Singleplayer saw big upgrades: Firstly, a decent **python chess engine** that playes chess was added.
- You can play against **stockfish chess engine** (see https://stockfishchess.org). My-PyChess will act as an interface to give you this singleplayer mode.
- En-Passant in chess, undo move, screenflip and sounds were added.
- Lots of changes made to GUI.

Click [here](CHANGELOG.md) to see full changelog.

## Online Gameplay
- You can self-host the My-PyChess online server(read more [here](onlinehowto.txt) ).
- Apart from that, I have launched a public My-PyChess server for PUBLIC BETA TESTING, which ANYONE in the world can connect to.

- Caveat: The server is on a IPv6-only network. This means that your network MUST SUPPORT IPv6 to connect. If your network does not support IPv6, try any other internet network. In my experience, many mobile networks are supporting IPv6 technology, so try mobile network tethering/hotspots.

### How to test wether your network supports IPv6
- A reliable way to test would be to enter "ipv6.google.com" in the browser window. If google pops up, then IPv6 is working for you.

## How to customise the game using preferences
- In the main game menu, click preferences.
- Hover over each name to know more about them.

1. Sounds: When True, sounds are enabled.
2. Flip: When True, it shows the chess board from the perspective of the player who is playing, otherwise shows a constant board with white side at the bottom.
3. SlideShow: When True, it will show a slideshow of images on main menu.
4. Show moves: When True, it will show all legal move options for a selected piece during gameplay.
5. Undo: When True, it allows users to undo.
6. Show Clock: When True, it shows a clock in multiplayer chess mode when timer is disabled, clock displayes total time elapsed since start.

- You can also open res/preferences.txt and edit the file to your preferences

## What's Next
- I prefer to work on this app locally, I do not commit to github for every change I make. I only commit when a new version is available or when I update any readme, etc.
- For the next version (v3.3), I plan to release less of new features and focus on the GUI along with any bugfixes or performance improvements. Mainly because v3.2 has come with a lot of code refactoring, new features and changes to the backend, and GUI hasn't recieved much attention by me in this release.


CHANGELOG.md
# Changelog
This is the basic changelog for My-PyChess. All dates are in dd/mm/yyyy format.

## What's new in Version 3.2 (2/9/2020)
- Added a Back-Button, to go back to the previous menu. In older versions, the quit button was used for this purpose, but from now on a dedicated button to go back is there in the top-right corner. The quit button will now be used only to exit the app.
- Added a game timer to multiplayer mode, with a new menu to setup timer.
- Fixed bugs and made many additions and improvements to the client-server in chess online gamemode.
- Made optimisations to core chess module and gui module.
- Added a chess hotwto.
- Minor improvements to game sounds and textbox.
- Upgraded preference menu and made the loadgame interface more robust.
- Several other minor changes and improvements made.

## What was new in Version 3.1 (14/6/2020)
- Added basic sounds (beta).
- Added draw and resign option to online gameplay.
- Added a simple "About menu".
- Added compatibility for python v3.5 (server.py still needs python v3.6 to work)
- Made many changes to GUI throughout the app, including more prompt messages.
- Upgraded preferences.
- Fixed a small issue in move animations, added a new startgame animation.
- Fixed a bug where undo in singleplayer after game ended would give weird results.
- Fixed many bugs in online mode (server), made server robust to TCP packet loss among other things.

## What was new in Version 3.0 (23/5/2020)
- THIS WAS THE BIGGEST RELEASE OF MY-PYCHESS EVER RELEASED.
- The code came with MIT License instead of the GPL-3 which came with v2.2.
- The code was revamped, fixing minor bugs and MAJOR PERFORMANCE IMPROVEMENTS.
- On my PC, a chess match with v2.2 ran at 22 fps, while this upgrade allowed it to run at over 200 fps if fps was not constrained.
- Online play was added. This featured a ONLINE LOBBY and support upto 10 people to play chess(read more in ref/online.txt)
- Singleplayer saw BIG UPGRADES: A Good Menu, ability to play with a decent PYTHON CHESS ENGINE.
- Now, you can play against STOCKFISH CHESS ENGINE. My-PyChess will act as a GUI interface to give you this singleplayer mode. 
- Included with this release, a good menu for stockfish and stockfish install/configure menu to help with installing and configuring stockfish with My-PyChess.
- Code was made more easy to understand with code comments throughout the code.
- EnPassant, Screen flip and Undo moves option were added.
- UI for loadGame menu was revamped, UI for preference menu was simplified.
- Save/load games with a new way of storing gamedata.

## What was new in Version 2.3 (4/2/2020)
- This version never made it to github, I dropped it while working on it and started to focus on v3.0
- This had included a few minor GUI changes.

## What was new in Version 2.2 (22/1/2020)
- Minor performance upgrades.
- Further upgraded preferences.
- Upgraded loadgame menu (with delete games option).
- Minor changes made to the interface.
- While saving a game, it informed under what name the game was being saved.

## What was new in Version 2.1 (11/1/2020)
- The upgrade optimised the game a bit, fixing a bug.
- When you clicked a piece, it would show all the available spaces it could go.
- The upgrade added a new preference menu where you could customise some game features.
- The upgrade added background animations on the home menu.

## What was new in Version 2.0 (7/1/2020)
- Revamped the Home Menu so that it looks better.
- Revamped and code, optimised it.
- Fixed Bugs.
- Made some changes that ensure that it can work on all platforms.
- Added Save/load game features.
- Added highly basic SinglePlayer (random move generator)

# Changelog for older versions

Below, I will mention about the initial development of this app. At this state, the game was unstable - It lacked many features, had performance issues and lot, lot of bugs. Most of the versions given below have not made it to github.

## What was new in Version 1.5 (26/11/2019)
- added choice for pawn promotion
- added piece animations
- added some popup messages
- upgraded home menu
- fixed bugs

## What was new in Version 1.4 (9/11/2019)
- Added a home menu
- Added placeholders for future upgrades
- Basic compliance with pep8

## What was new in Version 1.3 (2/11/2019)
- coding optimisations
- Checkmate and stalemate report feature added
- fixed bugs

## What was new in Version 1.2 (24/10/2019)
- Improved move validation, stopping king from moving into attacked areas.
- fixed bugs

## What was new in Version 1.1 (21/10/2019)
- added castling, pawn-promotion
- This version fixed a few bugs, but introduced lot more bugs

## What was new in Version 1.0 (19/10/2019)
- Added Move validation for highly basic chess moves

## What was new in Version 0.3 (17/10/2019)
- Added a few more basic utility functions.
- Fixed a few bugs.
- Minor changes made to code.

## What was new in Version 0.2 (16/10/2019)
- Added GUI for pieces, implemented basic piece logic.

## What was new in Version 0.1 (15/10/2019)
- Added GUI for the chess board


server.py
"""
This file is a part of My-PyChess application.
To run the online server, run this script.

For more information, see onlinehowto.txt

IMPORTANT NOTE:
    Server.py needs atleast Python v3.6 to work.
"""

import queue
import random
import socket
import threading
import time
from urllib.request import urlopen

# These are constants that can be modified by users. Default settings
# are given. Do not change if you do not know what you are doing.
LOG = False
IPV6 = False

#=====================================================
#        DO NOT MODIFY ANYTHING BELOW THIS!!
#=====================================================

# Define other constants
VERSION = "v3.2.0"
PORT = 26104
START_TIME = time.perf_counter()
LOGFILENAME = time.asctime().replace(" ", "_").replace(":", "-")

# Initialise a few global variables
busyPpl = set()
end = False
lock = False
logQ = queue.Queue()
players = []
total = totalsuccess = 0

# Function to convert string to int. Doesnt raise errors, returns None instead.
def makeInt(num):
    try:
        return int(num)
    except ValueError:
        return None

# A function to display elapsed time in desired format.
def getTime():
    sec = round(time.perf_counter() - START_TIME)
    minutes, sec = divmod(sec, 60)
    hours, minutes = divmod(minutes, 60)
    days, hours = divmod(hours, 24)
    return f"{days} days, {hours} hours, {minutes} minutes, {sec} seconds"

# A function to get IP address. It can give public IP or private.
def getIp(public):
    if public:
        try:
            ip = urlopen("https://api64.ipify.org").read().decode()
        except:
            ip = "127.0.0.1"
            
    else:  
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
            try:
                s.connect(('10.255.255.255', 1))
                ip = s.getsockname()[0]
            except:
                ip = '127.0.0.1'
    return ip

# A function to Log/Print text. Used instead of print()
def log(data, key=None, adminput=False):
    global logQ
    if adminput:
        text = ""
    elif key is None:
        text = "SERVER: "
    else:
        text = f"Player{key}: "
    
    if data is not None:
        text += data
        if not adminput:
            print(text)
    
        if LOG:
            logQ.put(time.asctime() + ": " + text + "\n")
    else:
        logQ.put(None)
        
# Used instead of sock.recv(), because it returns the decoded message, handles
# TCP packet loss, timeout and other useful stuff
def read(sock, timeout=None):
    try:
        sock.settimeout(timeout)
        msg = sock.recv(8).decode("utf-8").strip()

    except:
        msg = "quit"

    if msg:
        return msg
    return "quit"
 
# A function to message the server, this is used instead of socket.send()
# beacause it buffers the message, handles packet loss and does not raise
# exception if message could not be sent
def write(sock, msg):
    if msg:
        buffedmsg = msg + (" " * (8 - len(msg)))
        try:
            sock.sendall(buffedmsg.encode("utf-8"))
        except:
            pass

# Generates a random four digit number that is unique
def genKey():
    key = random.randint(1000, 9999)
    for player in players:
        if player[1] == key:
            return genKey()
    return key

# Given a players key, returns the sock object for the player
# Returns None if player does not exist
def getByKey(key):
    for player in players:
        if player[1] == makeInt(key):
            return player[0]

# Makes the player(s) busy, ie puts the player's key in a list of busy people
def mkBusy(*keys):
    global busyPpl
    for key in keys:
        busyPpl.add(makeInt(key))

# Makes the player(s) active, ie removes the player's key from the list of busy
def rmBusy(*keys):
    global busyPpl
    for key in keys:
        busyPpl.discard(makeInt(key))

# This simple function handles the chess match. Returns after game ended.
# Returns True if player got disconnected during the match, false otherwise.
def game(sock1, sock2):
    while True:
        msg = read(sock1)
        write(sock2, msg)
        if msg == "quit":
            return True

        elif msg in ["draw", "resign", "end"]:
            return False

# It handles every player's communiction with server.
def player(sock, key):
    while True:
        msg = read(sock)
        if msg == "quit":
            return

        elif msg == "pStat":
            log("Made request for players Stats.", key)
            latestplayers = list(players)
            latestbusy = list(busyPpl)
            
            if 0 < len(latestplayers) < 11:
                write(sock, "enum" + str(len(latestplayers) - 1))
                for _, i in latestplayers:
                    if i != key:
                        if i in latestbusy:
                            write(sock, str(i) + "b")
                        else:
                            write(sock, str(i) + "a")

        elif msg.startswith("rg"):
            log(f"Made request to play with Player{msg[2:]}", key)
            oSock = getByKey(msg[2:])
            if oSock is not None:
                if makeInt(msg[2:]) not in busyPpl:
                    mkBusy(key, msg[2:])
                    write(oSock, "gr" + str(key))
                    
                    write(sock, "msgOk")
                    newMsg = read(sock)
                    if newMsg == "ready":
                        log(f"Player{key} is in a game as white")
                        if game(sock, oSock):
                            return
                        else:
                            log(f"Player{key} finished the game")
                    
                    elif newMsg == "quit":
                        write(oSock, "quit")
                        return
                    
                    rmBusy(key)

                else:
                    log(f"Player{key} requested busy player")
                    write(sock, "errPBusy")
            else:
                log(f"Player{key} Sent invalid key")
                write(sock, "errKey")

        elif msg.startswith("gmOk"):
            log(f"Accepted Player{msg[4:]} request", key)
            oSock = getByKey(msg[4:])
            write(oSock, "start")
            log(f"Player{key} is in a game as black")
            if game(sock, oSock):
                return
            else:
                log(f"Player{key} finished the game")
                rmBusy(key)

        elif msg.startswith("gmNo"):
            log(f"Rejected Player{msg[4:]} request", key)
            write(getByKey(msg[4:]), "nostart")
            rmBusy(key)

# A thread to log all the texts. Flush from logQ.
def logThread():
    global logQ
    while True:
        time.sleep(1)
        with open("SERVER_LOG_" + LOGFILENAME + ".txt", "a") as f:
            while not logQ.empty(): 
                data = logQ.get()
                if data is None:
                    return
                else:
                    f.write(data)

# This is a Thread that runs in background to remove disconnected people
def kickDisconnectedThread():
    global players
    while True:
        time.sleep(10)
        for sock, key in players:
            try:
                ret = sock.send(b"........")
            except:
                ret = 0
                
            if ret > 0:
                cntr = 0
                diff = 8
                while True:
                    cntr += 1
                    if cntr == 8:
                        ret = 0
                        break
                        
                    if ret == diff:
                        break  
                    diff -= ret
                    
                    try:
                        ret = sock.send(b"." * diff)
                    except:
                        ret = 0
                        break         
                        
            if ret == 0:
                log(f"Player{key} got disconnected, removing from player list")
                try:
                    players.remove((sock, key))
                except:
                    pass
            
# This is a Thread that runs in background to collect user input commands
def adminThread():
    global end, lock
    while True:
        msg = input().strip()
        log(msg, adminput=True)

        if msg == "report":
            log(f"{len(players)} players are online right now,")
            log(f"{len(players) - len(busyPpl)} are active.")
            log(f"{total} connections attempted, {totalsuccess} were successful")
            log(f"Server is running {threading.active_count()} threads.")
            log(f"Time elapsed since last reboot: {getTime()}")
            if players:
                log("LIST OF PLAYERS:")
                for cnt, (_, player) in enumerate(players):
                    if player not in busyPpl:
                        log(f" {cnt+1}. Player{player}, Status: Active")
                    else:
                        log(f" {cnt+1}. Player{player}, Status: Busy")

        elif msg == "mypublicip":
            log("Determining public IP, please wait....")
            PUBIP = getIp(public=True)
            if PUBIP == "127.0.0.1":
                log("An error occurred while determining IP")
                
            else:
                log(f"This machine has a public IP address {PUBIP}")
                
        elif msg == "lock":
            if lock:
                log("Aldready in locked state")
            else:
                lock = True
                log("Locked server, no one can join now.")
                
        elif msg == "unlock":
            if lock:
                lock = False
                log("Unlocked server, all can join now.")
            else:
                log("Aldready in unlocked state.")

        elif msg.startswith("kick "):
            for k in msg[5:].split():
                sock = getByKey(k)
                if sock is not None:
                    write(sock, "close")
                    log(f"Kicking player{k}")
                else:
                    log(f"Player{k} does not exist")

        elif msg == "kickall":
            log("Attempting to kick everyone.")
            latestplayers = list(players)
            for sock, _ in latestplayers:
                write(sock, "close")

        elif msg == "quit":
            lock = True
            log("Attempting to kick everyone.")
            latestplayers = list(players)
            for sock, _ in latestplayers:
                write(sock, "close")
                
            log("Exiting application - Bye")
            log(None)
            
            end = True
            if IPV6:
                with socket.socket(socket.AF_INET6, socket.SOCK_STREAM) as s:
                    s.connect(("::1", PORT, 0, 0))
            else:
                with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                    s.connect(("127.0.0.1", PORT))
            return

        else:
            log(f"Invalid command entered ('{msg}').")
            log("See 'onlinehowto.txt' for help on how to use the commands.")

# Does the initial checks and lets players in.
def initPlayerThread(sock):
    global players, total, totalsuccess
    log("New client is attempting to connect.")
    total += 1
    
    if read(sock, 3) != "PyChess":
        log("Client sent invalid header, closing connection.")
        write(sock, "errVer")

    elif read(sock, 3) != VERSION:
        log("Client sent invalid version info, closing connection.")
        write(sock, "errVer")
    
    elif len(players) >= 10:
        log("Server is busy, closing new connections.")
        write(sock, "errBusy")
    
    elif lock:
        log("SERVER: Server is locked, closing connection.")
        write(sock, "errLock")
        
    else:
        totalsuccess += 1
        key = genKey()
        log(f"Connection Successful, assigned key - {key}")
        players.append((sock, key))
        
        write(sock, "key" + str(key))
        player(sock, key)
        write(sock, "close")
        log(f"Player{key} has Quit")
        
        try:
            players.remove((sock, key))
        except:
            pass
        rmBusy(key)
    sock.close()

# Initialize the main socket
log(f"Welcome to My-Pychess Server, {VERSION}\n")
log("INITIALIZING...")

if IPV6:
    log("IPv6 is enabled. This is NOT the default configuration.")
    
    mainSock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
    mainSock.bind(("::", PORT, 0, 0))
else:
    log("Starting server with IPv4 (default) configuration.")
    mainSock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    mainSock.bind(("0.0.0.0", PORT))
    
    IP = getIp(public=False)
    if IP == "127.0.0.1":
        log("This machine does not appear to be connected to a network.")
        log("With this limitation, you can only serve the clients ")
        log("who are on THIS machine. Use IP address 127.0.0.1\n")

    else:
        log(f"This machine has a local IP address - {IP}")
        log("USE THIS IP IF THE CLIENT IS ON THE SAME NETWORK.")
        log("For more info, read file 'onlinehowto.txt'\n")
    
mainSock.listen(16)
log("Successfully Started.")
log(f"Accepting connections on port {PORT}\n")

threading.Thread(target=adminThread).start()
threading.Thread(target=kickDisconnectedThread, daemon=True).start()
if LOG:
    log("Logging is enabled. Starting to log all output")
    threading.Thread(target=logThread).start()

while True:
    s, _ = mainSock.accept()
    if end:
        break

    threading.Thread(target=initPlayerThread, args=(s,), daemon=True).start()
mainSock.close()
